### PR TITLE
[SPARK-35841][SQL] Casting string to decimal type doesn't work if the…

### PR DESCRIPTION
**Pull request by dgd-contributor**
… sum of the digits is greater than 38
### What changes were proposed in this pull request?
Since Spark 3.1.1, NULL is returned when casting a string with many decimal places to a decimal type. If the sum of the digits before and after the decimal point is less than 39, a value is returned. From 39 digits, however, NULL is returned.
This worked until Spark 3.0.X.

Code to reproduce:

A string with 2 decimal places in front of the decimal point and 37 decimal places after the decimal point returns null

```
val data = Seq(
      "28.9259999999999983799625624669715762138",
      "28.925999999999998379962562466971576213",
      "2.9259999999999983799625624669715762138"
      )
val df = data.toDF("num")
df.withColumn("numConverted", col("num").cast("decimal(38, 5)")).show()
```

before this pull request, the result is
+----------------------+---------------+
|                 num          |numConverted|
+----------------------+---------------+
|28.92599999999999...|                  null|
|28.92599999999999...|         28.92600|
|2.925999999999998...|           2.92600|
+----------------------+---------------+

the correct result should be
+----------------------+---------------+
|                 num          |numConverted|
+----------------------+---------------+
|28.92599999999999...|         28.92600|
|28.92599999999999...|         28.92600|
|2.925999999999998...|           2.92600|
+----------------------+---------------+

The problem occur since https://issues.apache.org/jira/browse/SPARK-32706, it because the fast fail is checking precision length, which should only check the whole number part length of the input value, not the precision length

### Why are the changes needed?
correctness

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
test added

### CODE REVIEW

**Comment by dgd-contributor** [comment]
@cloud-fan @maropu can you review this code, since two of you had review pr #29731

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 1]
```diff
@@ -587,12 +587,8 @@ object Decimal {
     }
   }
 
-  private def calculatePrecision(bigDecimal: JavaBigDecimal): Int = {
-    if (bigDecimal.scale < 0) {
+  private def calculateWholeNumberPartLength(bigDecimal: JavaBigDecimal): Int = {
       bigDecimal.precision - bigDecimal.scale
```
[the remark this answers is not in the document]
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:591]: yes, as precision is the total number of digits (number of digit of a) and scale is the exponent (x) in 0.a * e ^x

[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:591]: then how about we rename it to `numIntegralDigits`?

[the remark this answers is not in the document]
Comment (dgd-contributor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala:591]: okay

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

**Review by HyukjinKwon — COMMENTED** [review]
lgtm2

**Pull request closed without merging**